agent/consul/multilimiter/multilimiter.go

```diff
@@ -0,0 +1,217 @@
+package multilimiter
+
+import (
+	"bytes"
+	"context"
+	radix "github.com/hashicorp/go-immutable-radix"
+	"golang.org/x/time/rate"
+	"sync/atomic"
+	"time"
+)
+
+var _ RateLimiter = &MultiLimiter{}
+
+const separator = "♣"
+
+func makeKey(keys ...[]byte) KeyType {
+	return bytes.Join(keys, []byte(separator))
+}
+
+func Key(prefix, key []byte) KeyType {
+	return makeKey(prefix, key)
+}
+
+// RateLimiter is the interface implemented by MultiLimiter
+type RateLimiter interface {
+	Run(ctx context.Context)
+	Allow(entity LimitedEntity) bool
+	UpdateConfig(c LimiterConfig, prefix []byte)
+}
+
+// MultiLimiter implement RateLimiter interface and represent a set of rate limiters
+// specific to different LimitedEntities and queried by a LimitedEntities.Key()
+type MultiLimiter struct {
+	limiters        *atomic.Pointer[radix.Tree]
+	limitersConfigs *atomic.Pointer[radix.Tree]
+	defaultConfig   *atomic.Pointer[Config]
+}
+
+type KeyType = []byte
+
+// LimitedEntity is an interface used by MultiLimiter.Allow to determine
+// which rate limiter to use to allow the request
+type LimitedEntity interface {
+	Key() KeyType
+}
+
+// Limiter define a limiter to be part of the MultiLimiter structure
+type Limiter struct {
+	limiter    *rate.Limiter
+	lastAccess atomic.Int64
+}
+
+// LimiterConfig is a Limiter configuration
+type LimiterConfig struct {
+	Rate  rate.Limit
+	Burst int
+}
+
+// Config is a MultiLimiter configuration
+type Config struct {
+	LimiterConfig
+	ReconcileCheckLimit    time.Duration
+	ReconcileCheckInterval time.Duration
+}
+
+// UpdateConfig will update the MultiLimiter Config
+// which will cascade to all the Limiter(s) LimiterConfig
+func (m *MultiLimiter) UpdateConfig(c LimiterConfig, prefix []byte) {
+	if prefix == nil {
+		prefix = []byte("")
+	}
+	configs := m.limitersConfigs.Load()
+	newConfigs, _, _ := configs.Insert(prefix, &c)
+	m.limitersConfigs.Store(newConfigs)
```

Comment:
Just leaving a note so we don't forget to add a lock here, to protect against concurrent calls to `UpdateConfig` overwriting each other's changes 🔒